interval/search_tree.go

```diff
@@ -178,3 +267,75 @@ func (st *MultiValueSearchTree[V, T]) IsEmpty() bool {
 
 	return st.root == nil
 }
+
+// GobEncode encodes the tree (compatible with [encoding/gob]).
+func (st *MultiValueSearchTree[V, T]) GobEncode() ([]byte, error) {
+	st.mu.RLock()
+	defer st.mu.RUnlock()
+
+	var b bytes.Buffer
+	enc := gob.NewEncoder(&b)
+
+	if err := enc.Encode(st.typeName()); err != nil {
+		return nil, err
+	}
+
+	if err := enc.Encode(st.config.allowIntervalPoint); err != nil {
+		return nil, err
+	}
+
+	if st.root != nil {
+		if err := enc.Encode(st.root); err != nil {
+			return nil, err
+		}
+	}
+
+	return b.Bytes(), nil
+}
+
+// GobDecode decodes the tree (compatible with [encoding/gob]).
+func (st *MultiValueSearchTree[V, T]) GobDecode(data []byte) error {
+	st.mu.Lock()
+	defer st.mu.Unlock()
+
+	b := bytes.NewBuffer(data)
+	enc := gob.NewDecoder(b)
+
+	var typeName string
+	wantTypeName := st.typeName()
+
+	if err := enc.Decode(&typeName); err != nil {
+		return err
+	}
+
+	if typeName != wantTypeName {
+		return TypeMismatchError{from: typeName, to: wantTypeName}
+	}
+
+	if err := enc.Decode(&st.config.allowIntervalPoint); err != nil {
+		return err
+	}
+
+	if err := enc.Decode(&st.root); err != nil {
+		if err != io.EOF {
+			return err
+		}
+
+		// An EOF error implies that the root
+		// wasn't encoded because it was nil
+		st.root = nil
+	}
+
+	return nil
+}
+
+func (st *MultiValueSearchTree[V, T]) typeName() string {
+	var v [0]V
+	var t [0]T
+
+	return fmt.Sprintf(
+		"MultiValueSearchTree[%v, %v]",
+		reflect.TypeOf(v).Elem().Name(),
+		reflect.TypeOf(t).Elem().Name(),
+	)
+}
```

Comment:
I see you point about encoding the generic types of the tree, but in this case I think this proposal goes against one of the main advantages of the [gob values](https://go.dev/blog/gob#values): _flexibility_. For example, it should not be an error when encoding a `MultiValueSearchTree[string, int]`, and decoding it to `MultiValueSearchTree[string, int8]`, because gob values aren't about types. When encoding an `int`, its value is transmitted as an unsized, variable-length integer; so as long as the receiver side decodes that into an arbitrary integer type in which the value fits in, everything should work.

Bottom line, I wouldn't restrict this by the generic types, only the tree type. Does that make sense?

Reply:
It feels inconsistent to check the tree type but not its generic types (one could argue the same thing about flexibility for the former case). But sure, I can change it.

Reply:
Glad you made that point. I just suggested to check for the tree types because they are not compatible: while `SearchTree` reads the values from the `val` field of `node` type, the `MultiValueSearchTree` reads values from the `vals` field. If it wasn't for that, I would totally agree with you that it would be inconsistent to test for the tree type but not for the generic types. Does that make sense?

Reply:
Just an FYI, using different generic types will also (obviously) provoke incompatibility issues.

e.g. `wrong type (string) for received field interval[string,int].Start`

So we have a tradeoff of flexibility vs misuse-verification.

Reply:
Done: [Don't check generic types](https://github.com/rdleal/intervalst/pull/9/commits/a6248274b6749e64e891a7b020de734d3f66487f)

Reply:
Yes! Gob decoding will help us in this regard. If users try to encode a (non-empty) `MultiValueSearchTree[string, string]` into a `MultiValueSearchTree[string, int]`, they will receive an error as you correctly exemplified. Which is expected and well documented in the [encoding/gob package](https://pkg.go.dev/encoding/gob#hdr-Types_and_Values).
I see your point about the tradeoff of flexibility vs misuse-verification, and I think it's worth for users to lose the flexibility of encoding a `SearchTree` and decoding it into a `MultiValueSearchTree` in order to make this package hard to misuse, because by design, both types are incompatible.
Thank you for the changes, btw.